Refactoring dockerapi and preparation for Python3

christianbur opened the pull request:
I wanted to rewrite the dockerapi to python3, but unfortunately there is a bug in the method 'exec_run' of docker-py https://github.com/mailcow/mailcow-dockerized/pull/2297 (in my opinion), for which I can't find a workaround at the moment. The PR contains my changes so far, I think the code is better readable. As far as I could test (not all methods are used by Mailcow) there were no bugs, but this should be checked again. When I have solved the problem with 'exec_run' I will create a new PR with Pyton3.

christianbur committed:
Update Dockerfile

christianbur committed:
Update server.py

andryyy commented:
Hm, I don't really want to merge a Py2 rewrite that **may** contain new bugs. :/ I confirm your code is better to read, yes! But I'm not sure we should merge a rewrite. Let's wait for a Py3 rewrite then.

Could you share your problem with exec_run, too? :) Perhaps someone knows a solution.

christianbur commented:
It's ok to wait for Python3, I just didn't know if there are any other activities on dockerapi at the moment. So somebody could continue to work on my previous changes. I had linked the problem already https://github.com/mailcow/mailcow-dockerized/pull/2297 

> use socket from exec_run in python3 (python2 is working) -- 'SocketIO' object has no attribute 'sendall' docker/docker-py#2255

christianbur commented:
This is the docker file for py3, the server.py script should run except for the two methods `def container_post__exec__rspamd__worker_password `and `def container_post__exec__system__mysql_upgrade`. Maybe someone has an idea.

```
FROM alpine:3.9
LABEL maintainer "Andre Peters <andre.peters@servercow.de>"

WORKDIR /app

RUN apk add -U --no-cache python3 tzdata openssl \
        python3-dev gcc musl-dev openssl-dev libffi-dev\
  && pip3 install --upgrade pip \
  && pip3 install --upgrade docker flask flask-restful pyOpenSSL \
  && apk del python3-dev gcc musl-dev openssl-dev libffi-dev

COPY server.py /

CMD ["python3", "-u", "/server.py"]
```

andryyy commented:
I don't have much time, but I'm curious as hell. :-D

christianbur commented:
it looks like someone found a solution to the exec_run problem, I'll take a look at it this weekend

christianbur commented:
So I did it :-) 
But I didn't know how to extend this pull request, I didn't want to create a new one (or do I really need to make a new PR?). Now someone could test, I didn't notice any more mistakes. 


the dockerapi-mailcow container has included the file `worker-controller-password.inc`
, i don't see any sense in that at the moment. Therefore I change the file in the method `container_post__exec__rspamd__worker_password `directly via the docker api. 

```
   dockerapi-mailcow:
      volumes:
         - ./data/conf/rspamd/override.d/worker-controller-password.inc:/access.inc:rw
```


Dockerfile 
```
FROM alpine:3.9
LABEL maintainer "Andre Peters <andre.peters@servercow.de>"

WORKDIR /app

RUN apk add -U --no-cache python3 tzdata openssl \
        python3-dev gcc musl-dev openssl-dev libffi-dev\
  && python3 -m pip install --upgrade pip \
  && python3 -m pip install --upgrade docker flask flask-restful pyOpenSSL \
  && apk del python3-dev gcc musl-dev openssl-dev libffi-dev

COPY server.py /

CMD ["python3", "-u", "/server.py"]

```

server.py
```
from flask import Flask
from flask_restful import Resource, Api
from flask import jsonify
from flask import Response
from flask import request
from threading import Thread
import docker
import uuid
import signal
import time
import os
import re
import sys
import ssl
import socket
import subprocess
import traceback

docker_client = docker.DockerClient(base_url='unix://var/run/docker.sock', version='auto')
app = Flask(__name__)
api = Api(app)

class containers_get(Resource):
  def get(self):
    containers = {}
    try:
      for container in docker_client.containers.list(all=True):
        containers.update({container.attrs['Id']: container.attrs})
      return containers
    except Exception as e:
      return jsonify(type='danger', msg=str(e))

class container_get(Resource):
  def get(self, container_id):
    if container_id and container_id.isalnum():
      try:
        for container in docker_client.containers.list(all=True, filters={"id": container_id}):
          return container.attrs
      except Exception as e:
          return jsonify(type='danger', msg=str(e))
    else:
      return jsonify(type='danger', msg='no or invalid id defined')

class container_post(Resource):
  def post(self, container_id, post_action):
    if container_id and container_id.isalnum() and post_action:
      try:
        """Dispatch container_post api call"""
        if post_action == 'exec':
          if not request.json or not 'cmd' in request.json:
            return jsonify(type='danger', msg='cmd is missing')
          if not request.json or not 'task' in request.json:
            return jsonify(type='danger', msg='task is missing')

          api_call_method_name = '__'.join(['container_post', str(post_action), str(request.json['cmd']), str(request.json['task']) ])
        else:
          api_call_method_name = '__'.join(['container_post', str(post_action) ])

        api_call_method = getattr(self, api_call_method_name, lambda container_id: jsonify(type='danger', msg='container_post - unknown api call'))


        print("api call: %s, container_id: %s" % (api_call_method_name, container_id))
        return api_call_method(container_id)
      except Exception as e:
        print("error - container_post: %s" % str(e))  
        return jsonify(type='danger', msg=str(e))

    else:
      return jsonify(type='danger', msg='invalid container id or missing action')        


  # api call: container_post - post_action: stop
  def container_post__stop(self, container_id):
    for container in docker_client.containers.list(all=True, filters={"id": container_id}):
      container.stop()
    return jsonify(type='success', msg='command completed successfully')


  # api call: container_post - post_action: start
  def container_post__start(self, container_id):
    for container in docker_client.containers.list(all=True, filters={"id": container_id}):
      container.start()
    return jsonify(type='success', msg='command completed successfully')


  # api call: container_post - post_action: restart
  def container_post__restart(self, container_id):
    for container in docker_client.containers.list(all=True, filters={"id": container_id}):
      container.restart()
    return jsonify(type='success', msg='command completed successfully')


  # api call: container_post - post_action: top
  def container_post__top(self, container_id):
    for container in docker_client.containers.list(all=True, filters={"id": container_id}):
      return jsonify(type='success', msg=container.top())


  # api call: container_post - post_action: stats
  def container_post__stats(self, container_id):
    for container in docker_client.containers.list(all=True, filters={"id": container_id}):
      for stat in container.stats(decode=True, stream=True):
        return jsonify(type='success', msg=stat )


  # api call: container_post - post_action: exec - cmd: mailq - task: delete
  def container_post__exec__mailq__delete(self, container_id):
    if 'items' in request.json:  
      r = re.compile("^[0-9a-fA-F]+$")
      filtered_qids = filter(r.match, request.json['items'])
      if filtered_qids:
        flagged_qids = ['-d %s' % i for i in filtered_qids]
        sanitized_string = str(' '.join(flagged_qids));

        for container in docker_client.containers.list(filters={"id": container_id}):
          postsuper_r = container.exec_run(["/bin/bash", "-c", "/usr/sbin/postsuper " + sanitized_string])
          return exec_run_handler('generic', postsuper_r)


  # api call: container_post - post_action: exec - cmd: mailq - task: hold
  def container_post__exec__mailq__hold(self, container_id):
    if 'items' in request.json:  
      r = re.compile("^[0-9a-fA-F]+$")
      filtered_qids = filter(r.match, request.json['items'])
      if filtered_qids:
        flagged_qids = ['-h %s' % i for i in filtered_qids]
        sanitized_string = str(' '.join(flagged_qids));

        for container in docker_client.containers.list(filters={"id": container_id}):
          postsuper_r = container.exec_run(["/bin/bash", "-c", "/usr/sbin/postsuper " + sanitized_string])
          return exec_run_handler('generic', postsuper_r)


   # api call: container_post - post_action: exec - cmd: mailq - task: unhold
  def container_post__exec__mailq__unhold(self, container_id):
    if 'items' in request.json:  
      r = re.compile("^[0-9a-fA-F]+$")
      filtered_qids = filter(r.match, request.json['items'])
      if filtered_qids:
        flagged_qids = ['-H %s' % i for i in filtered_qids]
        sanitized_string = str(' '.join(flagged_qids));

        for container in docker_client.containers.list(filters={"id": container_id}):
          postsuper_r = container.exec_run(["/bin/bash", "-c", "/usr/sbin/postsuper " + sanitized_string])
          return exec_run_handler('generic', postsuper_r)


  # api call: container_post - post_action: exec - cmd: mailq - task: deliver
  def container_post__exec__mailq__deliver(self, container_id):
    if 'items' in request.json:  
      r = re.compile("^[0-9a-fA-F]+$")
      filtered_qids = filter(r.match, request.json['items'])
      if filtered_qids:
        flagged_qids = ['-i %s' % i for i in filtered_qids]

        for container in docker_client.containers.list(filters={"id": container_id}):
          for i in flagged_qids:
            postqueue_r = container.exec_run(["/bin/bash", "-c", "/usr/sbin/postqueue " + i], user='postfix')
            # todo: check each exit code
          return jsonify(type='success', msg=str("Scheduled immediate delivery"))


  # api call: container_post - post_action: exec - cmd: mailq - task: list
  def container_post__exec__mailq__list(self, container_id):
    for container in docker_client.containers.list(filters={"id": container_id}):
      mailq_return = container.exec_run(["/usr/sbin/postqueue", "-j"], user='postfix')
      return exec_run_handler('utf8_text_only', mailq_return)


  # api call: container_post - post_action: exec - cmd: mailq - task: flush
  def container_post__exec__mailq__flush(self, container_id):
    for container in docker_client.containers.list(filters={"id": container_id}):
      postqueue_r = container.exec_run(["/usr/sbin/postqueue", "-f"], user='postfix')
      return exec_run_handler('generic', postqueue_r)


  # api call: container_post - post_action: exec - cmd: mailq - task: super_delete
  def container_post__exec__mailq__super_delete(self, container_id):
    for container in docker_client.containers.list(filters={"id": container_id}):
      postsuper_r = container.exec_run(["/usr/sbin/postsuper", "-d", "ALL"])
      return exec_run_handler('generic', postsuper_r)


  # api call: container_post - post_action: exec - cmd: system - task: fts_rescan
  def container_post__exec__system__fts_rescan(self, container_id):
    if 'username' in request.json:
      for container in docker_client.containers.list(filters={"id": container_id}):
        rescan_return = container.exec_run(["/bin/bash", "-c", "/usr/local/bin/doveadm fts rescan -u '" + request.json['username'].replace("'", "'\\''") + "'"], user='vmail')
        if rescan_return.exit_code == 0:
          return jsonify(type='success', msg='fts_rescan: rescan triggered')
        else:
          return jsonify(type='warning', msg='fts_rescan error')

    if 'all' in request.json:
      for container in docker_client.containers.list(filters={"id": container_id}):
        rescan_return = container.exec_run(["/bin/bash", "-c", "/usr/local/bin/doveadm fts rescan -A"], user='vmail')
        if rescan_return.exit_code == 0:
          return jsonify(type='success', msg='fts_rescan: rescan triggered')
        else:
          return jsonify(type='warning', msg='fts_rescan error')


  # api call: container_post - post_action: exec - cmd: system - task: df
  def container_post__exec__system__df(self, container_id):
    if 'dir' in request.json:
      for container in docker_client.containers.list(filters={"id": container_id}):
        df_return = container.exec_run(["/bin/bash", "-c", "/bin/df -H '" + request.json['dir'].replace("'", "'\\''") + "' | /usr/bin/tail -n1 | /usr/bin/tr -s [:blank:] | /usr/bin/tr ' ' ','"], user='nobody')
        if df_return.exit_code == 0:
          return df_return.output.decode('utf-8').rstrip()
        else:
          return "0,0,0,0,0,0"


  # api call: container_post - post_action: exec - cmd: system - task: mysql_upgrade
  def container_post__exec__system__mysql_upgrade(self, container_id):
    for container in docker_client.containers.list(filters={"id": container_id}):
      cmd = "/usr/bin/mysql_upgrade -uroot -p'" + os.environ['DBROOT'].replace("'", "'\\''") + "'\n"
      print(cmd)
      cmd_response = exec_cmd_container(container, cmd, user='mysql')
      print(cmd_response)

      matched = False
      for line in cmd_response.split("\n"):
        if 'is already upgraded to' in line:
          matched = True
      if matched:
        return jsonify(type='success', msg='mysql_upgrade: already upgraded')
      else:
        container.restart()
        return jsonify(type='warning', msg='mysql_upgrade: upgrade was applied')



  # api call: container_post - post_action: exec - cmd: reload - task: dovecot
  def container_post__exec__reload__dovecot(self, container_id):
    for container in docker_client.containers.list(filters={"id": container_id}):
      reload_return = container.exec_run(["/bin/bash", "-c", "/usr/local/sbin/dovecot reload"])
      return exec_run_handler('generic', reload_return)


  # api call: container_post - post_action: exec - cmd: reload - task: postfix
  def container_post__exec__reload__postfix(self, container_id):
    for container in docker_client.containers.list(filters={"id": container_id}):
      reload_return = container.exec_run(["/bin/bash", "-c", "/usr/sbin/postfix reload"])
      return exec_run_handler('generic', reload_return)


  # api call: container_post - post_action: exec - cmd: reload - task: nginx
  def container_post__exec__reload__nginx(self, container_id):
    for container in docker_client.containers.list(filters={"id": container_id}):
      reload_return = container.exec_run(["/bin/sh", "-c", "/usr/sbin/nginx -s reload"])
      return exec_run_handler('generic', reload_return)


  # api call: container_post - post_action: exec - cmd: sieve - task: list
  def container_post__exec__sieve__list(self, container_id):
    if 'username' in request.json:
      for container in docker_client.containers.list(filters={"id": container_id}):
        sieve_return = container.exec_run(["/bin/bash", "-c", "/usr/local/bin/doveadm sieve list -u '" + request.json['username'].replace("'", "'\\''") + "'"])
        return exec_run_handler('utf8_text_only', sieve_return)


  # api call: container_post - post_action: exec - cmd: sieve - task: print
  def container_post__exec__sieve__print(self, container_id):
    if 'username' in request.json and 'script_name' in request.json:
      for container in docker_client.containers.list(filters={"id": container_id}):
        cmd = ["/bin/bash", "-c", "/usr/local/bin/doveadm sieve get -u '" + request.json['username'].replace("'", "'\\''") + "' '" + request.json['script_name'].replace("'", "'\\''") + "'"]  
        sieve_return = container.exec_run(cmd)
        return exec_run_handler('utf8_text_only', sieve_return)


  # api call: container_post - post_action: exec - cmd: maildir - task: cleanup
  def container_post_exec__maildir__cleanup(self, container_id):
    if 'maildir' in request.json:
      for container in docker_client.containers.list(filters={"id": container_id}):
        sane_name = re.sub(r'\W+', '', request.json['maildir'])
        cmd = ["/bin/bash", "-c", "if [[ -d '/var/vmail/" + request.json['maildir'].replace("'", "'\\''") + "' ]]; then /bin/mv '/var/vmail/" + request.json['maildir'].replace("'", "'\\''") + "' '/var/vmail/_garbage/" + str(int(time.time())) + "_" + sane_name + "'; fi"]
        maildir_cleanup = container.exec_run(cmd, user='vmail')
        return exec_run_handler('generic', maildir_cleanup)



  # api call: container_post - post_action: exec - cmd: rspamd - task: worker_password
  def container_post__exec__rspamd__worker_password(self, container_id):
    if 'raw' in request.json:
      for container in docker_client.containers.list(filters={"id": container_id}):
        cmd = "/usr/bin/rspamadm pw -e -p '" + request.json['raw'].replace("'", "'\\''") + "' 2> /dev/null"
        cmd_response = exec_cmd_container(container, cmd, user="_rspamd")
        
        matched = False
        for line in cmd_response.split("\n"):
          if '$2$' in line:
            matched = True
            hash = line.strip()
            hash_out = re.search('\$2\$.+$', hash).group(0)
            rspamd_passphrase_hash = re.sub('[^0-9a-zA-Z\$]+', '', hash_out.rstrip())

            rspamd_password_filename = "/etc/rspamd/override.d/worker-controller-password.inc"
            cmd = '''/bin/echo 'enable_password = "%s";' > %s && cat %s''' % (rspamd_passphrase_hash, rspamd_password_filename, rspamd_password_filename)
            cmd_response = exec_cmd_container(container, cmd, user="_rspamd")

            container.restart()
        if matched:
            return jsonify(type='success', msg='command completed successfully')
        else:
            return jsonify(type='danger', msg='command did not complete')
        

def exec_cmd_container(container, cmd, user, timeout=2, shell_cmd="/bin/bash"):

  def recv_socket_data(c_socket, timeout):
    c_socket.setblocking(0)
    total_data=[];
    data='';
    begin=time.time()
    while True:
      if total_data and time.time()-begin > timeout:
        break
      elif time.time()-begin > timeout*2:
        break
      try:
        data = c_socket.recv(8192)
        if data:
          total_data.append(data.decode('utf-8'))
          #change the beginning time for measurement
          begin=time.time()
        else:
          #sleep for sometime to indicate a gap
          time.sleep(0.1)
          break
      except:
        pass
    return ''.join(total_data)
    
  try :
    socket = container.exec_run([shell_cmd], stdin=True, socket=True, user=user).output._sock
    if not cmd.endswith("\n"):
      cmd = cmd + "\n"
    socket.send(cmd.encode('utf-8'))
    data = recv_socket_data(socket, timeout)
    socket.close()
    return data

  except Exception as e:
    print("error - exec_cmd_container: %s" % str(e))
    traceback.print_exc(file=sys.stdout)

def exec_run_handler(type, output):
  if type == 'generic':
    if output.exit_code == 0:
      return jsonify(type='success', msg='command completed successfully')
    else:
      return jsonify(type='danger', msg='command failed: ' + output.output.decode('utf-8'))
  if type == 'utf8_text_only':
    r = Response(response=output.output.decode('utf-8'), status=200, mimetype="text/plain")
    r.headers["Content-Type"] = "text/plain; charset=utf-8"
    return r

class GracefulKiller:
  kill_now = False
  def __init__(self):
    signal.signal(signal.SIGINT, self.exit_gracefully)
    signal.signal(signal.SIGTERM, self.exit_gracefully)

  def exit_gracefully(self, signum, frame):
    self.kill_now = True

def create_self_signed_cert():
    process = subprocess.Popen(
      "openssl req -x509 -newkey rsa:4096 -sha256 -days 3650 -nodes -keyout /app/dockerapi_key.pem -out /app/dockerapi_cert.pem -subj /CN=dockerapi/O=mailcow -addext subjectAltName=DNS:dockerapi".split(),
      stdout = subprocess.PIPE, stderr = subprocess.PIPE, shell=False
    )
    process.wait()

def startFlaskAPI():
  create_self_signed_cert()
  try:
    ctx = ssl.create_default_context(ssl.Purpose.CLIENT_AUTH)
    ctx.check_hostname = False
    ctx.load_cert_chain(certfile='/app/dockerapi_cert.pem', keyfile='/app/dockerapi_key.pem')
  except:
    print ("Cannot initialize TLS, retrying in 5s...")
    time.sleep(5)
  app.run(debug=False, host='0.0.0.0', port=443, threaded=True, ssl_context=ctx)

api.add_resource(containers_get, '/containers/json')
api.add_resource(container_get, '/containers/<string:container_id>/json')
api.add_resource(container_post, '/containers/<string:container_id>/<string:post_action>')

if __name__ == '__main__':
  api_thread = Thread(target=startFlaskAPI)
  api_thread.daemon = True
  api_thread.start()
  killer = GracefulKiller()
  while True:
    time.sleep(1)
    if killer.kill_now:
      break
  print ("Stopping dockerapi-mailcow")
```

andryyy commented:
👍 I will test it.

christianbur commented:
I still have an idea for improving the api, to use the methods you always have to get the "Docker_ID" first.
Wouldn't it make more sense to provide a "container_name" URL to simply enter the container name?
"container_name" determine the id itself and internally call "containers" so both api calls would be available. This would simplify the use of the api and save code in the other containers. 

https://github.com/mailcow/mailcow-dockerized/blob/9213d65c853d7705601fbd79d0ce3f79d64b5b40/data/Dockerfiles/watchdog/watchdog.sh#L84

https://github.com/mailcow/mailcow-dockerized/blob/2c71c689f69b5dee5697518628557b4409039fb1/data/Dockerfiles/dovecot/sa-rules.sh#L17

https://github.com/mailcow/mailcow-dockerized/blob/99a816430bfb3f3f10c0dba49ef1d9067c572a8b/data/Dockerfiles/acme/docker-entrypoint.sh#L49

https://github.com/mailcow/mailcow-dockerized/blob/24432b0d9dae1b394b01e22e7f7aba337fc22834/data/Dockerfiles/phpfpm/docker-entrypoint.sh#L28

Now:
```
 #current 
 local NGINX=($(curl --silent --insecure https://dockerapi/containers/json | jq -r '.[] | {name: .Config.Labels["com.docker.compose.service"], id: .Id}' | jq -rc 'select( .name | tostring | contains("nginx-mailcow")) | .id' | tr "\n" " "))

 NGINX_RELOAD_RET=$(curl -X POST --insecure https://dockerapi/containers/${NGINX}/exec -d '{"cmd":"reload", "task":"nginx"}' --silent -H 'Content-type: application/json' | jq -r .type)
 [[ ${NGINX_RELOAD_RET} != 'success' ]] && { echo "Could not reload Nginx, restarting container..."; restart_container ${NGINX} ; }

----------------------------------------------------------------------------------------------- 

 #new
 NGINX_RELOAD_RET=$(curl -X POST --insecure https://dockerapi/container_name/nginx-mailcow/exec -d '{"cmd":"reload", "task":"nginx"}' --silent -H 'Content-type: application/json' | jq -r .type)
 [[ ${NGINX_RELOAD_RET} != 'success' ]] && { echo "Could not reload Nginx, restarting container..."; restart_container nginx-mailcow ; }
```

andryyy commented:
I didn't do that to be more flexible with multiple containers of the same service and to individually decide how to handle each ID (not doing that anywhere as of today though :)). This could still be moved/controlled by dockerapi, that's true.
We could add a handle "exec_each" or "exec_first" to decide on which container this command is run on. But not just exec, it needs to be implemented everywhere and for all endpoints, I think. 

Plus jq is cheap to run.... :) edit: and probably faster than Python code.

andryyy commented:
container_post__exec__rspamd__worker_password does not really check if it succeeded to write to the file. It just checks if the generated passwd is ok (matched = true).

```
        if matched:
            return jsonify(type='success', msg='command completed successfully')
```

Can you change this and perhaps also check if there are other occurrences where checks are not reliable? :)

christianbur committed:
Update Dockerfile

christianbur committed:
Update server.py

christianbur committed:
Update server.py

christianbur committed:
Update server.py

christianbur commented:
changed
```
            if rspamd_passphrase_hash.startswith("$2$") and rspamd_passphrase_hash in cmd_response:
              container.restart()
              matched = True
```
Dockerfile https://github.com/mailcow/mailcow-dockerized/pull/2342/commits/2383dd4c05d9e66bccc690f31b389f663af0b7f2 (`--virtual build-dependencies `added)
server.py https://github.com/mailcow/mailcow-dockerized/pull/2342/commits/b88b7565d6eebb9e023ee0d4e87e7cf45192d92d

In the method "container_post__exec__mailq__deliver" you had left the comment "# todo: check each exit code", after that you could have a look.

andryyy commented:
Nice, thank you!

I will run just a few more tests.

christianbur commented:
I find it ugly that the python script must run as root. Unfortunately, I can't think of another possibility. The reason is that the socket /var/run/docker.sock is only usable for USER=root and GROUP=docker. The root UID is always 0 (inside the container and outside) the GID of docker is different on every system (999, 120 on my systems).  And I don't know any way to dynamically determine the GID of the docker GID on the host system in the dockerfile. 

```
server# ls -al /var/run/docker.sock
srw-rw---- 1 root docker 0 Mar  2 18:34 /var/run/docker.sock

server# ls -nal /var/run/docker.sock
srw-rw---- 1 0 999 0 Mar  2 18:34 /var/run/docker.sock

server2# ls -nal /var/run/docker.sock
srw-rw---- 1 0 120 0 Mar  2 18:24 /var/run/docker.sock
```

christianbur commented:
@andryyy Did you already have time to test it?

christianbur commented:
Hello @andryyy,

is my implementation already live?
when i make the following command on my three installations i see the python3 script
`docker exec -it mailcowdockerized_dockerapi-mailcow_1 cat /server.py`

But here nothing has changed in the script for four months? I don't understand that right now.
https://github.com/mailcow/mailcow-dockerized/blob/636c780fb1bd68660386a451f5e4163c02252a82/data/Dockerfiles/dockerapi/server.py#L1

```
# docker images |grep dockerapi
mailcow/dockerapi         1.27                7df7e26c9109        2 months ago        83.4MB
```

christianbur committed:
Update Dockerfile

christianbur committed:
Update Dockerfile

andryyy commented on data/Dockerfiles/dockerapi/Dockerfile:
```diff
@@ -1,11 +1,14 @@
-  && pip2 install --upgrade pip \
-  && pip2 install --upgrade docker==3.0.1 flask flask-restful pyOpenSSL \
-  && apk del python-dev py2-pip gcc
+WORKDIR /app
```
Is there a reason for this? :)

christianbur replied:
What do you mean exactly?
This was my first Pull Reqest, so I had some problems with Github :-/ 
I had to delete the old respoistory, the current files are now here 

https://github.com/christianbur/mailcow-dockerized/blob/dockerapi_v3/data/Dockerfiles/dockerapi/Dockerfile

https://github.com/christianbur/mailcow-dockerized/blob/dockerapi_v3/data/Dockerfiles/dockerapi/server.py

andryyy replied:
The WORKDIR. Not sure you really need it.

Can you create a new PR with the new files?

andryyy commented:
I just need to find more time to test it. One thing that's very important is the detection of a running init_db process in php-fpm. I need to add a sleep to the init db script and check wether it is detected correctly.

The mysql_upgrade stuff can also be simulated by replacing the upgrade binary with a small script to return different strings.

Anyway... thank you very much for this!! :)
